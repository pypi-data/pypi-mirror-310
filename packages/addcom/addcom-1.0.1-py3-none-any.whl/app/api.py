from openai import OpenAI
from rich import print
import os


def build_prompt_messages(content: str, context: str) -> list[dict]:
    """
    Constructs messages for the LLM
    """
    # Raise an error if no code was provided
    if not content.strip():
        raise ValueError(
            "It seems that the specified file is empty, please provide code to comment"
        )

    # Create system prompt
    system_prompt = (
        "You are a coding assistant. When provided with the contents of a code file, your task is to add appropriate comments "
        "to explain its functionality where necessary. Comments should follow best practices and be concise yet informative. "
        "You may also receive example code snippets as a reference for the desired comment style. Carefully review these examples "
        "paying special attention to how functions and complex logic are explained. Sample snippets will be prefixed with 'Example:'. "
        "Your response should include only the modified code with the added comments,  without any additional text, explanations,"
        " or changes to the existing code. Make sure not to wrap the code in the brackets."
        "Again, don't provide any additional text with the code!"
    )

    # Initialize message list with the system prompt
    messages = [{"role": "system", "content": system_prompt}]

    # Add context file's contents as previous interaction - few-shot learning
    if context is not None:
        messages.append({"role": "user", "content": f"Example:\n{context}"})
        messages.append(
            {
                "role": "assistant",
                "content": (
                    "Great! Please provide another example if you have one, "
                    "or share the source code you'd like me to add comments to."
                ),
            }
        )

    # Add the uncommented code, as the last user message
    messages.append({"role": "user", "content": content})

    return messages


def generate_comments(
    file_path: str,
    content: str,
    context: str,
    api_key: str,
    url: str,
    model: str,
    stream: bool,
) -> str:
    """
    Send the file content to the API endpoint to generate comments.
    Returns the code with generated comments.
    """
    # Use environment variable for API key if it was not provided
    api_key = api_key or os.getenv("ADDCOM_API_KEY")

    # Check if API key was successfully set
    if not api_key:
        raise RuntimeError("Error: API key must be provided to generate comments")

    # Use Groq API endpoint as default if base URL not provided
    base_url = url or "https://api.groq.com/openai/v1"

    # Use LLama3 model as default if model option not provided
    model = model or "llama3-8b-8192"

    # Initialize OpenAI client (default = Groq API endpoint)
    client = OpenAI(base_url=base_url, api_key=api_key)

    try:
        response = client.chat.completions.create(
            messages=build_prompt_messages(content, context), model=model, stream=stream
        )

        if not stream:
            return response.choices[0].message.content
        else:
            streamed_content = ""

            print(f"--- {file_path} with added comments ---\n\n")

            # Print each chunk to stdout as it arrives and accumulate content
            for chunk in response:
                chunk_content = chunk.choices[0].delta.content
                if chunk_content:
                    print(chunk_content, end="", flush=True)
                    streamed_content += chunk_content

            print("\n\n")

            return streamed_content

    except Exception as e:
        raise RuntimeError(
            f"Error occurred while trying to generate comments for the file '{file_path}': {e}"
        )
