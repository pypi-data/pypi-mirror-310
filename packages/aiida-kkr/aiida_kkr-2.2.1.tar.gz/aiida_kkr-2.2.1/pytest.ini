[pytest]
testpaths = aiida_kkr
