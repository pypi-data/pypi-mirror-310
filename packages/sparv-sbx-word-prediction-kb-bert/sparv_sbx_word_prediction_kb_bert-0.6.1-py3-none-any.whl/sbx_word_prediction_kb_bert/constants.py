PROJECT_NAME: str = "sbx_word_prediction_kb_bert"
