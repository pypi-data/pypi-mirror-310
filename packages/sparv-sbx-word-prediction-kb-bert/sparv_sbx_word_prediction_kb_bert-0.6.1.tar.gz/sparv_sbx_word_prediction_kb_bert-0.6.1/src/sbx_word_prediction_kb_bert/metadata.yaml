id: swe-sbx-word-prediction-kb-bert
name:
    swe: Ordpredikation med KBLab/bert-base-swedish-cased för svenska
    eng: Word prediction with KBLab/bert-base-swedish-cased for Swedish
short_description:
    swe: Annotering av ord-predikation för varje text i en text.
    eng: Word prediction annotations for each word in a text.
task: word-prediction
annotations:
    - <token>:sbx_word_prediction_kb_bert.word-prediction--kb-bert
example_output: |-
    ```xml
    <token word="Engelbert" word-prediction--kb-bert="|Jag:0.388|Vi:0.384|Han:0.082|De:0.031|Hon:0.022|" pos="PM">Engelbert</token>
    <token word="tar" word-prediction--kb-bert="|tar:0.541|tog:0.208|kör:0.157|körde:0.050|åker:0.004|" pos="VB">tar</token>
    <token word="Volvon" word-prediction--kb-bert="|tunnelbanan:0.275|oss:0.118|bussen:0.116|mig:0.100|bilen:0.099|" pos="PM">Volvon</token>
    <token word="till" word-prediction--kb-bert="|till:0.897|från:0.038|mot:0.028|på:0.009|förbi:0.007|" pos="PP">till</token>
    <token word="Tele2" word-prediction--kb-bert="|Friends:0.584|Stockholm:0.136|Globen:0.037|Djurgården:0.034|Stockholms:0.027|" pos="PM">Tele2</token>
    <token word="Arena" word-prediction--kb-bert="|arena:0.518|Arena:0.471|,:0.002|Globen:0.001|Stockholm:0.001|" pos="PM">Arena</token>
    ```
