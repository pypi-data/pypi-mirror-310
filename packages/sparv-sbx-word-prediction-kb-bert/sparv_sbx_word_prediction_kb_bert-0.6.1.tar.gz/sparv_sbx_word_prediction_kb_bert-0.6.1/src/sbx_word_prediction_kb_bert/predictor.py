from dataclasses import dataclass
from typing import Optional, Tuple

import torch
from sparv import api as sparv_api  # type: ignore [import-untyped]
from transformers import (  # type: ignore [import-untyped]
    BertForMaskedLM,
    BertTokenizer,
    FillMaskPipeline,
)

logger = sparv_api.get_logger(__name__)

SCORE_FORMATS = {
    1: ("{:.1f}", lambda s: s.endswith(".0")),
    2: ("{:.2f}", lambda s: s.endswith(".00")),
    3: ("{:.3f}", lambda s: s.endswith(".000")),
    4: ("{:.4f}", lambda s: s.endswith(".0000")),
    5: ("{:.5f}", lambda s: s.endswith(".00000")),
    6: ("{:.6f}", lambda s: s.endswith(".000000")),
    7: ("{:.7f}", lambda s: s.endswith(".0000000")),
    8: ("{:.8f}", lambda s: s.endswith(".00000000")),
    9: ("{:.9f}", lambda s: s.endswith(".000000000")),
    10: ("{:.10f}", lambda s: s.endswith(".0000000000")),
}


class TopKPredictor:
    def __init__(
        self,
        *,
        tokenizer: Optional[BertTokenizer] = None,
        model: Optional[BertForMaskedLM] = None,
        num_decimals: int = 3,
    ) -> None:
        self.tokenizer = tokenizer or self._default_tokenizer()
        self.model = model or self._default_model()
        self.num_decimals = num_decimals
        self.pipeline = FillMaskPipeline(model=self.model, tokenizer=self.tokenizer)

    @classmethod
    def _default_model(cls) -> BertForMaskedLM:
        if torch.cuda.is_available():
            logger.info("Using GPU (cuda)")
            dtype = torch.float16
        else:
            logger.warning("Using CPU, is cuda available?")
            dtype = torch.float32
        model = BertForMaskedLM.from_pretrained(
            MODELS["kb-bert"].model_name,
            revision=MODELS["kb-bert"].model_revision,
            torch_dtype=dtype,
            device_map=(
                "auto"
                if torch.cuda.is_available() and torch.cuda.device_count() > 1
                else None
            ),
        )
        if torch.cuda.is_available() and torch.cuda.device_count() == 1:
            model = model.cuda()  # type: ignore
        return model  # type: ignore

    @classmethod
    def _default_tokenizer(cls) -> BertTokenizer:
        tokenizer_name, tokenizer_revision = MODELS[
            "kb-bert"
        ].tokenizer_name_and_revision()

        return BertTokenizer.from_pretrained(
            tokenizer_name, revision=tokenizer_revision
        )

    def get_top_k_predictions(self, text: str, k: int = 5) -> str:
        tokenized_inputs = self.tokenizer(text)
        if len(tokenized_inputs["input_ids"]) <= 512:  # type: ignore
            return self._run_pipeline(text, k)
        if text.count("[MASK]") == 1:
            return self._run_pipeline_on_mask_context(text, k)
        raise RuntimeError(
            f"can't handle large input and multiple [MASK]: {len(tokenized_inputs['input_ids'])} tokens > 512 tokens"  # noqa: E501 # type: ignore
        )

    def _run_pipeline_on_mask_context(self, text, k):
        start, end = self.compute_context(text)
        text_with_mask = text[start:end]
        return self._run_pipeline(text_with_mask, k)

    def compute_context(self, text):
        mask = text.find("[MASK]")
        lower = text[(mask - 210) : (mask - 190)].find(" ")
        higher = text[(mask + 190) : (mask + 210)].find(" ")
        start = mask - 210 + lower
        end = mask + 190 + higher
        return max(start, 0), min(end, len(text))

    def _run_pipeline(self, text, k) -> str:
        if predictions := self.pipeline(text, top_k=k):
            collect_token_and_score = (
                (pred["token_str"], pred["score"])  # type: ignore
                for pred in predictions
            )
            score_format, score_pred = SCORE_FORMATS[self.num_decimals]
            format_scores = (
                (token, score_format.format(score))
                for token, score in collect_token_and_score
            )
            filter_out_zero_scores = (
                (token, score)
                for token, score in format_scores
                if not score_pred(score)
            )
            predictions_str = "|".join(
                f"{token}:{score}" for token, score in filter_out_zero_scores
            )

            return f"|{predictions_str}|" if predictions_str else "|"
        else:
            return "|"


@dataclass
class HuggingfaceModel:
    model_name: str
    model_revision: str
    tokenizer_name: Optional[str] = None
    tokenizer_revision: Optional[str] = None

    def tokenizer_name_and_revision(self) -> Tuple[str, str]:
        if tokenizer_name := self.tokenizer_name:
            return tokenizer_name, self.tokenizer_revision or "main"
        else:
            return self.model_name, self.model_revision


MODELS = {
    "kb-bert": HuggingfaceModel(
        model_name="KBLab/bert-base-swedish-cased",
        model_revision="c710fb8dff81abb11d704cd46a8a1e010b2b022c",
    )
}
