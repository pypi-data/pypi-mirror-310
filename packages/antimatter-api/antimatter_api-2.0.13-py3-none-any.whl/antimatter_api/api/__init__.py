# flake8: noqa

# import apis into api package
from antimatter_api.api.account_management_api import AccountManagementApi
from antimatter_api.api.authentication_api import AuthenticationApi
from antimatter_api.api.capsules_api import CapsulesApi
from antimatter_api.api.contexts_api import ContextsApi
from antimatter_api.api.encryption_api import EncryptionApi
from antimatter_api.api.general_api import GeneralApi
from antimatter_api.api.internal_api import InternalApi
from antimatter_api.api.keychain_api import KeychainApi
from antimatter_api.api.policy_api import PolicyApi

