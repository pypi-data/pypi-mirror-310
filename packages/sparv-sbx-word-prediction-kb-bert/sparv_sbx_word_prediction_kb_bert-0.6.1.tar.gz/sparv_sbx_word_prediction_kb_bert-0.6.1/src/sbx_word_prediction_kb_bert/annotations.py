from typing import Optional

from sparv import api as sparv_api  # type: ignore [import-untyped]
from sparv.api import (  # type: ignore [import-untyped]
    Annotation,
    Config,
    Output,
)

from sbx_word_prediction_kb_bert.constants import PROJECT_NAME
from sbx_word_prediction_kb_bert.predictor import TopKPredictor

logger = sparv_api.get_logger(__name__)
TOK_SEP = " "


def load_predictor(num_decimals_str: str) -> TopKPredictor:
    try:
        num_decimals = int(num_decimals_str)
    except ValueError as exc:
        raise sparv_api.SparvErrorMessage(
            f"'{PROJECT_NAME}.num_decimals' must contain an 'int' got: '{num_decimals_str}'"  # noqa: E501
        ) from exc

    return TopKPredictor(num_decimals=num_decimals)


@sparv_api.annotator(
    "Word prediction tagging with a masked Bert model",
    language=["swe"],
    preloader=load_predictor,
    preloader_params=["num_decimals_str"],
    preloader_target="predictor_preloaded",
)
def predict_words__kb_bert(
    out_prediction: Output = Output(
        f"<token>:{PROJECT_NAME}.word-prediction--kb-bert",
        description="Word predictions from masked BERT (format: '|<word>:<score>|...|)",
    ),
    word: Annotation = Annotation("<token:word>"),
    sentence: Annotation = Annotation("<sentence>"),
    num_predictions_str: str = Config(f"{PROJECT_NAME}.num_predictions"),
    num_decimals_str: str = Config(f"{PROJECT_NAME}.num_decimals"),
    predictor_preloaded: Optional[TopKPredictor] = None,
) -> None:
    logger.info("predict_words")
    try:
        num_predictions = int(num_predictions_str)
    except ValueError as exc:
        raise sparv_api.SparvErrorMessage(
            f"'{PROJECT_NAME}.num_predictions' must contain an 'int' got: '{num_predictions_str}'"  # noqa: E501
        ) from exc

    predictor = predictor_preloaded or load_predictor(num_decimals_str)

    sentences, _orphans = sentence.get_children(word)
    token_word = list(word.read())
    out_prediction_annotation = word.create_empty_attribute()

    run_word_prediction(
        predictor=predictor,
        num_predictions=num_predictions,
        sentences=sentences,
        token_word=token_word,
        out_prediction_annotations=out_prediction_annotation,
    )

    logger.info("writing annotations")
    out_prediction.write(out_prediction_annotation)


def run_word_prediction(
    predictor: TopKPredictor,
    num_predictions: int,
    sentences,
    token_word: list,
    out_prediction_annotations,
) -> None:
    logger.info("run_word_prediction")

    logger.progress(total=len(sentences))  # type: ignore
    for sent in sentences:
        logger.progress()  # type: ignore
        token_indices = list(sent)
        for token_index_to_mask in token_indices:
            sent_to_tag = TOK_SEP.join(
                (
                    "[MASK]"
                    if token_index == token_index_to_mask
                    else token_word[token_index]
                )
                for token_index in sent
            )

            predictions_scores = predictor.get_top_k_predictions(
                sent_to_tag, k=num_predictions
            )
            out_prediction_annotations[token_index_to_mask] = predictions_scores
