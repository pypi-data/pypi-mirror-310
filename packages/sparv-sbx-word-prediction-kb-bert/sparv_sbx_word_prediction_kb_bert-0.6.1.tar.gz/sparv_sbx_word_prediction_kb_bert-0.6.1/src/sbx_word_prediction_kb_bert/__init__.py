from sparv import api as sparv_api  # type: ignore [import-untyped]

from sbx_word_prediction_kb_bert.annotations import predict_words__kb_bert

__all__ = ["predict_words__kb_bert"]

__description__ = "Calculating word predictions by mask a word in a BERT model."


__config__ = [
    sparv_api.Config(
        "sbx_word_prediction_kb_bert.num_predictions",
        description="The number of predictions to list",
        default=5,
    ),
    sparv_api.Config(
        "sbx_word_prediction_kb_bert.num_decimals",
        description="The number of decimals to round the score to",
        default=3,
    ),
]

__version__ = "0.6.1"
