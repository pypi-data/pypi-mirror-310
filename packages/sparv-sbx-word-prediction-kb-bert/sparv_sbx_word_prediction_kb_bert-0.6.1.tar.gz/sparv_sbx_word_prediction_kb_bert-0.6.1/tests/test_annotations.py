import pytest
from sbx_word_prediction_kb_bert.annotations import (
    load_predictor,
    predict_words__kb_bert,
)
from sparv.api import SparvErrorMessage  # type: ignore [import-untyped]
from sparv_pipeline_testing import MemoryOutput, MockAnnotation


def test_load_predictor_not_int_raises() -> None:
    with pytest.raises(SparvErrorMessage):
        load_predictor("not an int")


def test_predict_words__kb_bert__bad_int_raises() -> None:
    with pytest.raises(SparvErrorMessage):
        predict_words__kb_bert(
            out_prediction=MemoryOutput(),
            word=MockAnnotation(),
            sentence=MockAnnotation(),
            num_predictions_str="not-an-int",
            num_decimals_str="2",
        )


def test_predict_words__kb_bert(snapshot) -> None:
    output: MemoryOutput = MemoryOutput()

    word = MockAnnotation(
        name="<token:word>", values=["Han", "åt", "glassen", "utanför", "kiosken", "."]
    )
    sentence = MockAnnotation(
        name="<sentence>", children={"<token:word>": [[0, 1, 2, 3, 4, 5]]}
    )

    predict_words__kb_bert(
        output, word, sentence, num_predictions_str="5", num_decimals_str="2"
    )

    assert output.values == snapshot
