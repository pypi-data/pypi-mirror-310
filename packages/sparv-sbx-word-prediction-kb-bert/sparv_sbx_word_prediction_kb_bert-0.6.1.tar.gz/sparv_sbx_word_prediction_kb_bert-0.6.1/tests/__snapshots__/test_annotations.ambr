# serializer version: 1
# name: test_predict_words__kb_bert
  list([
    '|Vi:0.25|Jag:0.18|De:0.13|Hon:0.09|Han:0.09|',
    '|köpte:0.18|sålde:0.15|hämtade:0.04|tog:0.04|tappade:0.04|',
    '|frukost:0.23|lunch:0.21|middag:0.09|korv:0.04|hamburgare:0.03|',
    '|i:0.92|från:0.03|utanför:0.02|vid:0.02|',
    '|fönstret:0.21|dörren:0.05|restaurangen:0.05|huset:0.04|kiosken:0.03|',
    '|.:0.99|?:0.01|',
  ])
# ---
